src/wallet/wallet.cpp

```diff
@@ -4872,20 +4883,95 @@ void CWallet::GetFilteredNotes(
     std::string address,
     int minDepth,
     bool ignoreSpent,
-    bool requireSpendingKey)
+    bool requireSpendingKey,
+    boost::optional<uint64_t> timestamp)
 {
     std::set<PaymentAddress> filterAddresses;
 
     if (address.length() > 0) {
         filterAddresses.insert(DecodePaymentAddress(address));
     }
 
-    GetFilteredNotes(sproutEntries, saplingEntries, filterAddresses, minDepth, INT_MAX, ignoreSpent, requireSpendingKey);
+    GetFilteredNotes(sproutEntries, saplingEntries, filterAddresses, minDepth, INT_MAX, ignoreSpent, requireSpendingKey, true, timestamp);
+}
+
+boost::iterator_range<NotesHelperIndex::by_timestamp_itr> CWallet::GetNotesByType(
+    uint256 paymentaddress_hash,
+    NoteType type,
+    boost::optional<uint64_t> timestamp)
+{
+    int countAlreadySkipped = 0;
+
+    auto instance = NotesHelperIndex::getInstance();
+    auto& index = instance->index.get<by_timestamp>();
+
+    uint64_t ts = 0;
+    if (timestamp) {
+        ts = *timestamp;
+    }
+    auto itr_start = index.upper_bound(std::make_tuple(type, paymentaddress_hash, ts));
+    auto itr_end = index.upper_bound(std::make_tuple(type, paymentaddress_hash));
```

Comment:
Here I'm no sure, can you doublecheck that this is correct?

I don't know how tuples are ordered in C++. In Python, to which I'm used to it goes like this:
```
>>> (1,2) > (1,2,1)
False
>>> (1,2,sys.maxint) > (1,2,1)
True
```

If the ordering is the same in C++ than the `itr_end` should have tuple like `(type, paymentaddress_hash, INT_MAX)`

Reply:
`s/INT_MAX/UINT64_MAX/`

Reply:
https://github.com/zcash/zcash/pull/4399/commits/a2584a676860a9cae73e67cbd0a6ac5a880f6687